IpmiFeaturePkg/Library/IpmiCommandLib/IpmiCommandLibNetFnApp.c

```diff
@@ -245,3 +246,81 @@ IpmiSendMessage (
              );
   return Status;
 }
+
+/**
+ * Function which returns the System Guid returned from the BMC
+ *
+ * @param SystemGuid Pointer the the buffer to fill with System Guid
+ * @return EFI_STATUS
+ */
+EFI_STATUS
+EFIAPI
+IpmiGetSystemGuid (
+  OUT EFI_GUID  *SystemGuid
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      DataSize;
+
+  //  IPMI_GET_DEVICE_GUID_RESPONSE is associted with Get Device Guid (NetFn App,  Cmd 0x08), but
+  //   the response from Get System Guid is the same format. Reusing the structure here.
+  IPMI_GET_SYSTEM_GUID_RESPONSE  GuidResponse;
+
+  Status = EFI_INVALID_PARAMETER;
+  if (SystemGuid != NULL) {
+    DataSize = sizeof (IPMI_GET_DEVICE_GUID_RESPONSE);
+
+    Status = IpmiSubmitCommand (
+               IPMI_NETFN_APP,
+               IPMI_APP_GET_SYSTEM_GUID,
+               NULL,
+               0,
+               (VOID *)&GuidResponse,
+               &DataSize
+               );
+    if (!EFI_ERROR (Status)) {
+      CopyMem ((VOID *)SystemGuid, (VOID *)(GuidResponse.Guid), sizeof (EFI_GUID));
+    }
+  }
+
+  return Status;
+}
+
+/**
+ * @brief Query the BMC for the information about the system interface
+ *
+ * @param Type The interface that is being queried (GET_SYSTEM_INTEFACE_INTERFACE_TYPE)
+ * @param GetSystemInterfaceResponse Pointer to a buffer for returning the response data
+ * @param ResponseDataSize Size of the buffer for returning the response data
+ * @return EFI_STATUS
+ */
+EFI_STATUS
+EFIAPI
+IpmiGetSystemInterfaceCapabilities (
+  IN GET_SYSTEM_INTEFACE_INTERFACE_TYPE                 Type,
+  IN OUT IPMI_GET_SYSTEM_INTERFACE_CAPABILITY_RESPONSE  *GetSystemInterfaceResponse,
+  IN OUT UINT32                                         *ResponseDataSize
+  )
+{
+  EFI_STATUS  Status;
+
+  IPMI_GET_SYSTEM_INTERFACE_CAPABILITY_REQUEST  RequestData;
+
+  RequestData.SystemInterfaceType = Type;
+  RequestData.Reserved            = 0;
+
+  Status = EFI_INVALID_PARAMETER;
+
+  if ((GetSystemInterfaceResponse != NULL) &&  (ResponseDataSize != NULL) && (*ResponseDataSize >= sizeof (IPMI_GET_SYSTEM_INTERFACE_CAPABILITY_RESPONSE))) {

```

Comment:
Looking at the spec, I think the response size can vary from type to type. Specifically, KCS & SMIC only take 4 bytes but SSIF takes 5. Do we want to consider it valid for them to only pass in 4 bytes? If so then the current check would check against the greatest size in the enum.

Reply:
I was hoping the union of structures would get around that, but you bring up a good point. 

When IpmiSubmitCommand is called, is there a problem if the buffer that is passed in is too large? i.e. does it require exact buffer sizing, or will it just fill the buffer with the amount of bytes the command returns?

I.e. if we pass the 5 byte buffer size, but was are calling for KCS or SMIC, will it be fine handling it?

The check on the response buffer would also need to be modified if this is a problem, as it is using the union as the minimum size requirement....

Reply:
It should be fine if the buffer is oversized, it will only use what it needs and should update the pointer size to update the written size. So we can easily just require the max possible size (of 5 bytes) and simplify it. We could even remove the size param at that point since its already a structure pointer anyways.

Reply:
My hope is the caller will use the enum to allow for the difference in the return types.